### DISCUSSION TITLE
supervisor.step() returns too early

### DISCUSSION THREAD

**Discussion by KelvinYang0320** [aidudezzz/deepbots#76] [Q&A] [answered]
**<code>supervisor.step()</code> returns too early**
I want the robotic arm to rotate its joints -0.05, 0, or 0.05 in each step, but the robot hasn't done its action, and directly returns current values from its position sensors. Then the supervisor sends next action to the robot.
However, the experiment still works and seems good.

To make the robotic arm rotate its joints precisely, I let the robot send a specific message which denotes that it is still moving, and let the supervisor send robot a special action number, say -1, which denotes that supervisor is not sending an action to the robot but asking whether the robot stops moving.

The above solution can solve the problem, but I am looking for some more efficient ways.
Thanks!

**Comment by tsampazk** [comment]
Hey @KelvinYang0320,

This issue sounds peculiar. Can you share controllers or point me to code that reproduces this behaviour?

**Reply by KelvinYang0320** [reply]
[My deepbots-panda issue2 repo](https://github.com/KelvinYang0320/deepbots-panda/tree/issue-2-supervisor_step_returns_too_early)
* [supervisor.step()](https://github.com/KelvinYang0320/deepbots-panda/blob/cc78376cac03b0911e1e5b5ffe7f099abeddd24a/Panda_RL/controllers/supervisorController/supervisorController.py#L135)
* ![擷取](https://user-images.githubusercontent.com/49781698/108487882-b466b900-72da-11eb-81da-b5fd07cc6f15.PNG)
* my solution:
    *  [supervisorController](https://github.com/KelvinYang0320/deepbots-panda/blob/cc78376cac03b0911e1e5b5ffe7f099abeddd24a/Panda_RL/controllers/supervisorController/supervisorController.py#L136)
    * [robotController](https://github.com/KelvinYang0320/deepbots-panda/blob/cc78376cac03b0911e1e5b5ffe7f099abeddd24a/Panda_RL/controllers/robotController/robotController.py#L42)

Still lots of dirty code, sorry.
And thank you for your reply!

**Reply by tsampazk** [reply]
> Still lots of dirty code, sorry.

No worries, i will check it out and get back to you!

**Comment by tsampazk** [comment, marked as answer]
Hello @KelvinYang0320, i took a look at the code and the example and - correct me if i am wrong - i think that a robot action takes more than one step to complete. As a consequence the supervisor has to wait for the previous action to complete before getting the new observations and sending back the new action.

Webots motors [controlled by setPosition run an internal PID controller](https://cyberbotics.com/doc/reference/motor#position-control) to reach the position you asked for. This can take multiple steps. The solution you implemented seems perfectly fine to me. I think you are forced to implement such a scheme (or anything similar really) to wait for the action to complete.
 
I would suggest that you use the [RobotSupervisor scheme](https://github.com/aidudezzz/deepbots#combined-robot-supervisor-scheme), you can see an example [here](https://github.com/tsampazk/deepworlds/tree/cartpole-disc-robsuper/examples/cartpole/cartpole_discrete/controllers/robot_supervisor_manager). Note that the linked example uses the latest dev release of deepbots which you can find [here](https://test.pypi.org/project/deepbots/).
 
With the robot-supervisor scheme we use the robot node as a supervisor, without needing any emitter-receiver communication, and all needed functions are implemented in one controller. Moreover, the example uses the new reset procedure that simplifies reset and makes it far-far faster. Switching from the emitter-receiver to the robot-supervisor scheme is more a matter of copy-pasting and deleting some stuff, no big deal really (i have done it myself in the example i linked). The new scheme will make your code far simpler and more efficient. 

_I think that by switching to the new scheme your issue will be solved by default._
 
 Feel free to ask any other questions and i'll be happy to help.

**Reply by KelvinYang0320** [reply]
Thank you for your advice. I will go through your deepworlds example and give it a try.

**Reply by KelvinYang0320** [reply]
Hello @tsampazk,
using the robot-supervisor scheme really make training far faster in your [cartpole example](https://github.com/tsampazk/deepworlds/tree/cartpole-disc-robsuper/examples/cartpole/cartpole_discrete/controllers/robot_supervisor_manager), that is great!

However, if I still want to make the robotic arm rotate its joints precisely, with the robot-supervisor scheme, I still need the process to wait for the previous action to complete.
> To make the robotic arm rotate its joints precisely, I let the robot send a specific message which denotes that it is still moving, and let the supervisor send robot a special action number, say -1, which denotes that supervisor is not sending an action to the robot but asking whether the robot stops moving.

[deepbots-panda with deepbots 0.1.3.dev2](https://github.com/KelvinYang0320/deepbots-panda/tree/issue-3-switch_to_the_new_scheme_deepbots0_1_3_dev2/Panda_RL/controllers/robot_supervisor_manager)
Click the button: run the simulation as fast as possible, both using [deepbots 0.1.3.dev2](https://test.pypi.org/project/deepbots/) with pretrained  DDPG models from my [IKPY-based-RL-ENV repo](https://github.com/KelvinYang0320/IKPY-based-RL-ENV/tree/plane_reaching)
(handshaking means the process (code:[link1](https://github.com/KelvinYang0320/deepbots-panda/blob/ac669a387c61b6b6e8f1765bfd651f2ff4e94fa6/Panda_RL/controllers/robot_supervisor_manager/robot_supervisor_ddpg.py#L82) [link2](https://github.com/KelvinYang0320/deepbots-panda/blob/ac669a387c61b6b6e8f1765bfd651f2ff4e94fa6/Panda_RL/controllers/robot_supervisor_manager/DDPG_runner.py#L38)) to wait for the previous action to complete)
| No handshaking (ignore any error) | handshaking (error<0.001)| 
| -------- | -------- |
|![](https://i.imgur.com/CR6XDwd.gif)   | ![](https://i.imgur.com/GFRb42T.gif)    | 

If I train panda with this handshaking process, it will take more time to complete training.
Besides, since the robotSupervisor scheme is more efficient, it seems to casue more handshaking times than using emitter-receiver.

Thank you!

**Reply by tsampazk** [reply]
You are right, the problem is not solved, my bad! At least you don't have to do the handshaking process through the emitters and receivers, which simplifies things a bit. 😁

This is an inherent issue of the way discrete actions work. For the PPO discrete case, the deltaAngle you use in combination with the desired motor velocity that you set, mean that the motors will not reach the target position of +- deltaAngle that the action dictates in a single step. You can read up [here](https://cyberbotics.com/doc/reference/motor#motor-functions), in the description where it reads "change the parameters of the PID-controller" it describes more or less what you are doing, waiting for each movement to complete.

To describe what i think is happening in the gifs you attached, is that on the left one where the position of the motor is not controlled with the handshaking process, the agent continuously repeats certain actions that bring the end effector closer to the target and thus many +deltaAngles are added in the motor position continuously, so lets say for target motor position TP starting at 0.0 in timestep T=0, real motor position RP starting also at 0.0, deltaAngle=0.005 we have:

| TP | RP | T |
| ----------- | ----------- | ----------- |
| 0.000 | 0.000 | 0 |
| 0.005 | 0.001 | 1 |
| 0.010 | 0.002 | 2 |
| 0.015 | 0.004 | 3 |
| 0.020 | 0.008 | 4 |
|... | ...| ...|

Note that the real position (RP) is controlled by the PID controller and i put the values in arbirtrarily, just following a trend. So the target position increases, the real position lags behind and so the velocity is increased more and more by the PID controller to catch up, causing a slow start, really high velocity around the middle of the "path" and rapid slow down at the end as seen in the gif. On the other gif on the right, the PID controller is given time to reach each new position and thus the increase/decrease in velocity is done in small steps and is not visible giving the illusion of precision.

I think that your precision is defined by the deltaAngle you use and not the completion of each action before doing a new one. I drew a diagram below that illustrates the idea in an extreme case where it shows the last step of reaching a small target with a large deltaAngle:

![image](https://user-images.githubusercontent.com/27914645/109290373-32bde080-7830-11eb-9742-8156471e3312.png)

Nevertheless, your solution seems to work great. Another solution is to use a much much smaller deltaAngle and make sure the target position is reached in a single step, which will indeed increase the precision. Another would be to lower the desired motorVelocity by a lot (e,g, 0.5 instead of 2.5), so as to not get these big spikes in the velocity of the arm. Both these solutions make the arm movement much slower. To be honest, i am not familiar with the intricacies of this particular problem so i cannot offer an opinion on what approach is better. With large targets, a big deltaAngle might be enough and your solution does the trick. 

As a side note, one problem i spotted is that when the arm hits the floor and tries to move through it, the target position is never reached and the simulation hangs, never going through the handshake process, as seen in the screenshot below.

![image](https://user-images.githubusercontent.com/27914645/109288124-22583680-782d-11eb-8b9e-8558d056b633.png)

On the other hand, the continuous DDPG approach can solve the issue altogether by allowing the agent to control the seven motor velocities a directly, instead of the positions. Then you don't need a deltaAngle at all. I tried this out and as expected it is much more difficult for the agent to learn to control the arm via the velocities, probably needs more time and maybe even a more elaborate reward function. 

This could also be done with the discrete PPO algorithm, which can increase/decrease the velocity instead of the positions, and when a max velocity is reached you can clip the value and keep it steady. This requires again a delta value but now its deltaVelocity, which i think enables the agent to control the position much more precisely. Imagine that it puts a motor at max velocity by increasing it by deltaVelocity and when it reaches the target it zeroes it out in a single step, using a stop action. This way you can make sure you have both velocity control (by assigning a max value to clip) and precise position control.

Sorry for the really long reply, it's a really nice problem to think about and try to solve!

**Reply by KelvinYang0320** [reply]
@tsampazk Thank you very much for your detailed advice! 
I will try to solve these problems.

**Reply by KelvinYang0320** [reply]
@tsampazk Thank you very much for helping me!
I have fixed the problem that you mentioned, thanks for pointing out the problem.
> As a side note, one problem i spotted is that when the arm hits the floor and tries to move through it, the target position is never reached and the simulation hangs, never going through the handshake process, as seen in the screenshot below.

I have trained the DDPG agent from scratch without using the pretrained model from my [IKPY-based-RL-ENV](https://github.com/KelvinYang0320/IKPY-based-RL-ENV/tree/plane_reaching), and it seems to converge after more than 10000 episodes which takes more than one day. It seems to work well!
For more details, please refer to my [deepworlds fork](https://github.com/KelvinYang0320/deepworlds/tree/dev/examples/panda).

**Reply by tsampazk** [reply]
Happy to help @KelvinYang0320. Whenever you feel ready, open a pull request so we can include the problem in the deepworlds examples, which will make a great addition to the repository!

**Reply by KelvinYang0320** [reply]
@tsampazk I have done refactor but found that I forgot to save my models during training. 
Now I have to train DDPG agent again to ensure everything goes well.
It might take about two days, sorry:cry:

**Reply by tsampazk** [reply]
No worries, we are not including trained models in the deepworlds repository anyway, at least not yet. The code that is used for training is enough!

**Reply by KelvinYang0320** [reply]
Ok, thank you @tsampazk .
I will open a pull request for this code that is used for training.
If there are new issues after two days, I will fix them and open another new pull request.
